tests/migrations/test_executor.py

```diff
@@ -670,6 +672,15 @@ def test_migrate_marks_replacement_unapplied(self):
                 executor.recorder.applied_migrations(),
             )
 
+    @override_settings(MIGRATION_MODULES={'migrations': 'migrations.test_migrations_squashed'})
+    def test_partially_applied_replacement_raises_informative_error(self):
+        recorder = MigrationRecorder(connection)
+        recorder.record_applied('migrations', '0001_initial')
+        executor = MigrationExecutor(connection)
+        msg = 'Migration migrations.0001_squashed_0002 is incompletely applied.'
```

Comment:
Having somehow never encountered this problem (that I can recall), and based on the detail in the ticket, I'm unsure of the answer so I'm just going to ask a question and hope it's not too dumb ;)

Is this message useful enough? Once the user finds themselves in this situation and sees that message, will they have an understanding of what needs to be done? 
- If yes, all's well and good.
- If no ... is there any extra guidance or information (eg: internal migration state data) that could be plucked to explain the next step, whatever that is?

Reply:
That's a good thought. Let me think for a day or two and see what I can come up with 👍🏻

Reply:
So we've got a user asking Django to migrate to a squashed migration that's in a confused partial state with some but not all of that squashed migration's replacements applied on the box in question. So yes, if you end up here, you'd probably be dying for more information.

I see a couple options:
 - just direct folks to `manage.py showmigrations --list` which should be enough to get the state of play
 - print out all the unapplied migrations preventing the user from accomplishing what they came here to accomplish

I just pushed a commit implementing option 2 but could be convinced of either. What do you think?

Reply:
Personally if it were happening to me, I'd be **thrilled** if it could tell me what needs applying, and if it couldn't and instead directed me to another command to run I'd _assume_ (and be disappointed) that the information wasn't available at the time the exception occurred. If it is there, I think it's prudent to respect the user's needs and attempt to surface the information.

That said, my gut feeling is that the exception output format isn't "right" (for a given value thereof). I can't recall many (any?) places where Django raises with an arg which is multiple lines. Those cases where there's multiple _things_ to enumerate tend to just get comma separated. That does leave the door open to super long lists though, if for whatever reason the partially applied list was big...

Reply:
> Maybe because it's related to running migrations, which are line oriented, it's OK?

That was my thought, but you have a point, it seems nonstandard. I'll remove it.

I was curious, so I looked and found at least [one instance](https://github.com/django/django/blob/ca9872905559026af82000e46cde6f7dedc897b6/tests/template_tests/syntax_tests/test_load.py#L54-L75) where newlines are used, but as you indicated, I wouldn't follow this example in most cases.